index.d.ts

```diff
@@ -211,10 +223,119 @@ export interface DailyRoomInfo {
+    | 'recording-stats'
+    | 'recording-error'
+    | 'recording-upload-completed'
+    | 'fullscreen'
+    | 'exited-fullscreen'
+    | 'live-stream-started'
+    | 'live-stream-stopped'
+    | 'track-started'
+    | 'track-stopped'
+  >;
+};
+
+export type DailyEventErrorObject = {
+  action: Extract<
+    DailyEvent,
+    'load-attempt-failed' | 'live-stream-error' | 'error'
+  >;
+  errorMsg: string;
+};
+
+export type DailyEventObjectParticipants = {
+  action: Extract<DailyEvent, 'joined-meeting'>;
+  participants: DailyParticipantsObject;
+};
+
+export type DailyEventObjectParticipant = {
+  action: Extract<
+    DailyEvent,
+    'participant-joined' | 'participant-updated' | 'participant-left'
+  >;
+  participant: DailyParticipant;
+};
+
+export type DailyEventObjectMouseEvent = {
+  action: Extract<
+    DailyEvent,
+    'click' | 'mousedown' | 'mouseup' | 'mouseover' | 'mousemove'
+  >;
+  event: MouseEvent;
+};
+
+export type DailyEventObjectTouchEvent = {
+  action: Extract<DailyEvent, 'touchstart' | 'touchmove' | 'touchend'>;
+  event: TouchEvent;
+};
+
+export type DailyEventObjectNetworkQualityEvent = {
+  action: Extract<DailyEvent, 'network-quality-change'>;
+  threshold: string;
+  quality: number;
+};
+
+export type NetworkConnectionType = 'signaling' | 'peer-to-peer' | 'sfu';
+
+export type DailyEventObjectNetworkEvent = {
+  action: Extract<DailyEvent, 'network-quality-change'>;
+  type: NetworkConnectionType;
+  event: string;
+  session_id?: string;
+  sfu_id?: string;
+};
+
+export type DailyEventObjectActiveSpeakerChange = {
+  action: Extract<DailyEvent, 'active-speaker-change'>;
+  activeSpeaker: {
+    peerId: string;
+  };
+};
+
+export type DailyEventObjectActiveSpeakerModeChange = {
+  action: Extract<DailyEvent, 'active-speaker-mode-change'>;
+  enabled: boolean;
+};
+
+export type DailyEventObjectAppMessage = {
+  action: Extract<DailyEvent, 'app-message'>;
+  data: any;
+};
+
+export type DailyEventObject<
+  T extends DailyEvent
+> = T extends DailyEventObjectAppMessage['action']
+  ? DailyEventObjectAppMessage
+  : T extends DailyEventObjectNoPayload['action']
+  ? DailyEventObjectNoPayload
+  : T extends DailyEventErrorObject['action']
+  ? DailyEventErrorObject
+  : T extends DailyEventObjectParticipants['action']
+  ? DailyEventObjectParticipants
+  : T extends DailyEventObjectParticipant['action']
+  ? DailyEventObjectParticipant
+  : T extends DailyEventObjectMouseEvent['action']
+  ? DailyEventObjectMouseEvent
+  : T extends DailyEventObjectTouchEvent['action']
+  ? DailyEventObjectTouchEvent
+  : T extends DailyEventObjectNetworkQualityEvent['action']
+  ? DailyEventObjectNetworkQualityEvent
+  : T extends DailyEventObjectNetworkEvent['action']
+  ? DailyEventObjectNetworkEvent
+  : T extends DailyEventObjectActiveSpeakerChange['action']
+  ? DailyEventObjectActiveSpeakerChange
+  : T extends DailyEventObjectActiveSpeakerModeChange['action']
+  ? DailyEventObjectActiveSpeakerModeChange
+  : any;
```

Comment:
I made this optionally generic... This looks gross in the type def but in practice it gets you this:

![Screen Shot 2020-07-07 at 7 20 59 PM](https://user-images.githubusercontent.com/13444851/86867170-35907900-c087-11ea-99b9-dd6429d5b3a4.png)

In other words, if you start an event handler and say `.on("participant-updated",` it knows what the event type is based on the first argument